### PR TITLE
Do not use PkLookUp if any pk cols are under NOT predicate

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
Fixes https://github.com/crate/crate/issues/15458 and https://github.com/crate/crate/issues/15395.

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/test/java/io/crate/analyze/where/EqualityExtractorTest.java [hunk 1]
```diff
@@ -95,6 +95,24 @@ public void testNoExtractOnNotEqualsOnSinglePk() {
...
+    @Test
+    public void test_no_pk_extraction_under_not_operator() {
+        List<List<Symbol>> matches = analyzeExactX(query("x != 1 or x = 1"));
+        assertThat(matches).isNull();
+
+        matches = analyzeExactX(query("x != 1 and x = 1"));
+        assertThat(matches).isNull();
+
+        matches = analyzeExactX(query("not(x != 1)"));
+        assertThat(matches).isNull();
```
Comment (jeeminso) [inline server/src/test/java/io/crate/analyze/where/EqualityExtractorTest.java:107]: I expect cases like this must be rewritten to `x = 1` elsewhere.

**Comment by jeeminso** [comment]
This is off track but current `EqualityExtractor` is difficult to understand, for example `x=1 or x is null`, `x is null` becomes `Literal.BOOLEAN_TRUE` but not sure if it is intentional.

I was thinking if we can improve on this by introducing `DomainFinder` which accepts a boolean query and a set of columns (primary keys) that returns the domain of each given columns to satisfy the query.

For example,
```
create table t (a int primary key, b int primary key);
select * from t where true; -- a=[int_min, int_max], b=[int_min, int_max]
select * from t where a = 1; -- a=[1], b=[int_min, int_max]
select * from t where a = 1 and b = 1; -- a=[1], b=[1]
select * from t where a = any([1,2,3]); -- a=[1,2,3], b=[int_min, int_max]
```
More complex example:
```
-- or
(a = 1 and b = 1) or (a = 2 and b = 2) =>  (a=[1], b=[1]) or (a=[2], b=[2]) => 2 doc-keys
a = 1 or (a = 2 and b = 2) => (a=[1], b=[int_min, int_max]) or (a=[2], b=[2]) => ~ 2*int_max number of doc-keys
```
```
-- and
(a = 1 and b = 1) and (a = 2 and b = 2) => (a=[1], b=[1]) and (a=[2], b=[2]) => a=[],b=[] => 0 doc-keys
a = 1 and (a = 2 and b = 2) => (a=[1], b=[int_min, int_max]) and (a=[2], b=[2]) => a=[],b=[2] => 0 doc-keys
```

**Comment by jeeminso** [comment]
Please review this draft @seut, @mkleen. I could also see some would prefer more fine grain logic.

**Comment by seut** [comment]
> This is off track but current EqualityExtractor is difficult to understand, for example x=1 or x is null, x is null becomes Literal.BOOLEAN_TRUE but not sure if it is intentional.

We already select the related records by the recorded DocKeys (primary keys). So this query should result (given by the OR operator) in collected pk values `[1, null]` and the related part of the where clause should be stripped/replaced by true.

File: server/src/test/java/io/crate/analyze/where/EqualityExtractorTest.java [hunk 2]
```diff
@@ -95,6 +95,36 @@ public void testNoExtractOnNotEqualsOnSinglePk() {
         assertThat(matches).isNull();
     }
 
+    @Test
+    public void test_no_extraction_if_pk_under_not_operator() {
+        List<List<Symbol>> matches = analyzeExactX(query("x != 1 or x = 1"));
+        assertThat(matches).isNull();
+
+        matches = analyzeExactX(query("x != 1 and x = 1"));
+        assertThat(matches).isNull();
```
[the remark this answers is not in the document]
Comment (jeeminso) [reply server/src/test/java/io/crate/analyze/where/EqualityExtractorTest.java]: Thank you, I moved them to `ColumnsUnderNotPredicateFinderTest` and created tests specific to the bug scenarios.

**Review by seut — APPROVED** [review]
👍 thx!

**Pull request merged**